### DISCUSSION TITLE
How to use the slots API in toktx?

### DISCUSSION THREAD

**Discussion by krispya** [donmccurdy/glTF-Transform#1161] [Q&A] [answered]
I am struggling to properly use the `slots` prop for `toktx`. The examples I found seem to be outdated. I want to compress the color texture with ETC1S and the rest with UASTC but nothing I do with regex seems to match a slot. Is there a list of what common slots are named, or what I am supposed to be matching against?

**Comment by donmccurdy** [comment, marked as answer]
Hi @krispya - are you using the CLI (with globs), or a script (with regex)?

The names of slots come directly from the glTF specification and its extensions. Examples:

- baseColorTexture
- normalTexture
- occlusionTexture
- metallicRoughnessTexture
- emissiveTexture
- ...

As found in:

- https://registry.khronos.org/glTF/specs/2.0/glTF-2.0.html#reference-material-pbrmetallicroughness
- https://github.com/KhronosGroup/glTF/blob/main/extensions/2.0/Khronos/KHR_materials_transmission/README.md#properties
- https://github.com/KhronosGroup/glTF/blob/main/extensions/2.0/Khronos/KHR_materials_clearcoat/README.md#clearcoat

More generally, I think it is the case that all current color slots include the words 'color', 'emissive', or 'diffuse', so this regex should be enough to select them: `/color|emissive|diffuse/i`. When given as a CLI argument, you'd need an equivalent glob instead of a regex.

**Reply by krispya** [reply]
Thanks. I am using the script with regex. I found this example inside the source code as well.

```js
const MICROMATCH_OPTIONS = { nocase: true, contains: true };
const slotsUASTC = micromatch.makeRe(
	'{normalTexture,occlusionTexture,metallicRoughnessTexture}',
	MICROMATCH_OPTIONS
);

await document.transform(
	toktx({ mode: Mode.UASTC, slots: slotsUASTC }),
	toktx({ mode: Mode.ETC1S })
);
```

**Reply by donmccurdy** [reply]
I should perhaps revisit that example – micromatch is certainly convenient. But it might have been better for me to write a matcher for _color_ textures to make them ETC1S first, and using UASTC for whatever is left. Rather than trying to match _data_ textures to make them UASTC (of which there are many others beyond those matched above), and making everything else ETC1S. But ultimately it's whatever works best for your content here -- no rule saying you can't use different modes in other ways!

**Reply by krispya** [reply]
Thanks, I also found `listTextureSlots` which helps me identify what I am trying to target. https://gltf-transform.dev/modules/functions/functions/listTextureSlots
